### PR TITLE
[SPARK-29052][DOCS][ML][PYTHON][CORE][R][SQL][SS] Create a Migration Guide tap in Spark documentation

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

Currently, there is no migration section for PySpark, SparkCore and Structured Streaming.
It is difficult for users to know what to do when they upgrade.

This PR proposes to create create a "Migration Guide" tap at Spark documentation.

Screen Shot 2019-09-11 at 7 02 05 PM

Screen Shot 2019-09-11 at 7 27 15 PM

This page will contain migration guides for Spark SQL, PySpark, SparkR, MLlib, Structured Streaming and Core. Basically it is a refactoring.

There are some new information added, which I will leave a comment inlined for easier review.

1. **MLlib**
  Merge [ml-guide.html#migration-guide](https://spark.apache.org/docs/latest/ml-guide.html#migration-guide) and [ml-migration-guides.html](https://spark.apache.org/docs/latest/ml-migration-guides.html)

    ```
    'docs/ml-guide.md'
            ↓ Merge new/old migration guides
    'docs/ml-migration-guide.md'
    ```

2. **PySpark**
  Extract PySpark specific items from https://spark.apache.org/docs/latest/sql-migration-guide-upgrade.html

    ```
    'docs/sql-migration-guide-upgrade.md'
           ↓ Extract PySpark specific items
    'docs/pyspark-migration-guide.md'
    ```

3. **SparkR**
  Move [sparkr.html#migration-guide](https://spark.apache.org/docs/latest/sparkr.html#migration-guide) into a separate file, and extract from [sql-migration-guide-upgrade.html](https://spark.apache.org/docs/latest/sql-migration-guide-upgrade.html)

    ```
    'docs/sparkr.md'                     'docs/sql-migration-guide-upgrade.md'
     Move migration guide section ↘     ↙ Extract SparkR specific items
                     docs/sparkr-migration-guide.md 
    ```

4. **Core**
  Newly created at `'docs/core-migration-guide.md'`. I skimmed resolved JIRAs at 3.0.0 and found some items to note.

5. **Structured Streaming**
  Newly created at `'docs/ss-migration-guide.md'`. I skimmed resolved JIRAs at 3.0.0 and found some items to note.
 
6. **SQL**
  Merged [sql-migration-guide-upgrade.html](https://spark.apache.org/docs/latest/sql-migration-guide-upgrade.html) and [sql-migration-guide-hive-compatibility.html](https://spark.apache.org/docs/latest/sql-migration-guide-hive-compatibility.html)
    ```
    'docs/sql-migration-guide-hive-compatibility.md'     'docs/sql-migration-guide-upgrade.md'
     Move Hive compatibility section ↘                   ↙ Left over after filtering PySpark and SparkR items
                                  'docs/sql-migration-guide.md'
    ```

### Why are the changes needed?

In order for users in production to effectively migrate to higher versions, and detect behaviour or breaking changes before upgrading and/or migrating.

### Does this PR introduce any user-facing change?
Yes, this changes Spark's documentation at https://spark.apache.org/docs/latest/index.html.

### How was this patch tested?

Manually build the doc. This can be verified as below:

```bash
cd docs
SKIP_API=1 jekyll build
open _site/index.html
```

### CODE REVIEW

File: docs/core-migration-guide.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+{:toc}
+
+## Upgrading from Core 2.4 to 3.0
+
+- In Spark 3.0, deprecated method `TaskContext.isRunningLocally` has been removed. Local execution was removed and it always has returned `false`.
+
+- In Spark 3.0, deprecated method `shuffleBytesWritten`, `shuffleWriteTime` and `shuffleRecordsWritten` in `ShuffleWriteMetrics` have been removed. Instead, use `bytesWritten`, `writeTime ` and `recordsWritten` respectively.
+
+- In Spark 3.0, deprecated method `AccumulableInfo.apply` have been removed because creating `AccumulableInfo` is disallowed.
+
```
Comment (HyukjinKwon) [inline docs/core-migration-guide.md]: cc @srowen . I added those items from https://github.com/apache/spark/commit/0025a8397f8723011917239fe47518457d4d6860 while I am adding this section.

File: docs/ml-migration-guide.md [hunk 2]
```diff
@@ -20,15 +19,80 @@ license: |
...
 
-## From 2.1 to 2.2
+Note that this migration guide describes the items specific to MLlib.
+Many items of SQL migration can be applied when migrating MLlib to higher versions for DataFrame-based APIs.
+Please refer [Migration Guide: SQL, Datasets and DataFrame](sql-migration-guide.html).
+
+## Upgrading from MLlib 2.4 to 3.0
+
+### Breaking changes
+{:.no_toc}
```
Comment (HyukjinKwon) [inline docs/ml-migration-guide.md]: `{:.no_toc}` is needed to exclude in the "Table of Contents". For example:

Without no_toc
Screen Shot 2019-09-11 at 7 30 08 PM

With no_toc
Screen Shot 2019-09-11 at 7 30 24 PM

**Comment by HyukjinKwon** [comment]
FYI /cc @gatorsmile @cloud-fan @BryanCutler @felixcheung @viirya

File: docs/core-migration-guide.md (hunk 1 above)
Comment (srowen) [reply docs/core-migration-guide.md]: There's probably more we can or should add, given the number of deprecations and removals in 3.0. Is your theory that this should stick to changes that require user code to change, and ones that aren't obvious? Like, a missing method is obvious. A behavior change may not be.

**Review by dongjoon-hyun — APPROVED** [review]
For me, this looks like a good start. We can add more.
BTW, which version of Jekyll do you use, @HyukjinKwon ?
In the `master` branch (and this PR), I saw misaligned tabs when I use Jekyll 4.0.

**Comment by dongjoon-hyun** [comment]
Ur, it seems to be not a Jekyll issue. When I use `spark-rm` docker image, it still seems to be broken. Maybe, something went wrong. cc @gatorsmile 
```
spark-rm@f327b886c883:/spark/docs$ jekyll --version
jekyll 3.8.6
spark-rm@f327b886c883:/spark/docs$ ruby --version
ruby 2.3.8p459 (2018-10-18 revision 65136) [x86_64-linux-gnu]
```

Screen Shot 2019-09-13 at 10 07 24 AM

File: docs/pyspark-migration-guide.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+Note that this migration guide describes the items specific to PySpark.
+Many items of SQL migration can be applied when migrating PySpark to higher versions.
+Please refer [Migration Guide: SQL, Datasets and DataFrame](sql-migration-guide.html).
+
+## Upgrading from PySpark 2.4 to 3.0
+
+  - Since Spark 3.0, PySpark requires a Pandas version of 0.23.2 or higher to use Pandas related functionality, such as `toPandas`, `createDataFrame` from Pandas DataFrame, etc.
+
+  - Since Spark 3.0, PySpark requires a PyArrow version of 0.12.1 or higher to use PyArrow related functionality, such as `pandas_udf`, `toPandas` and `createDataFrame` with "spark.sql.execution.arrow.enabled=true", etc.
```
Comment (viirya) [inline docs/pyspark-migration-guide.md]: "spark.sql.execution.arrow.enabled=true" => ``` `spark.sql.execution.arrow.enabled`=true ``` or ``` `spark.sql.execution.arrow.enabled` set to true ```?

File: docs/pyspark-migration-guide.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+  - In PySpark, `df.replace` does not allow to omit `value` when `to_replace` is not a dictionary. Previously, `value` could be omitted in the other cases and had `None` by default, which is counterintuitive and error-prone.
+
+## Upgrading from PySpark 1.4 to 1.5
+
+ - Resolution of strings to columns in Python now supports using dots (`.`) to qualify the column or
+   access nested values. For example `df['table.column.nestedField']`. However, this means that if
+   your column name contains any dots you must now escape them using backticks (e.g., ``table.`column.with.dots`.nested``).
+
+ - DataFrame.withColumn method in PySpark supports adding a new column or replacing existing columns of the same name.
```
Comment (viirya) [inline docs/pyspark-migration-guide.md]: `DataFrame.withColumn`

File: docs/_data/menu-sql.yaml [hunk 5]
```diff
@@ -64,15 +64,7 @@
...
-  url: sql-migration-guide.html
-  subitems:
-    - text: Spark SQL Upgrading Guide
-      url: sql-migration-guide-upgrade.html
-    - text: Compatibility with Apache Hive
-      url: sql-migration-guide-hive-compatibility.html
-    - text: SQL Reserved/Non-Reserved Keywords
-      url: sql-reserved-and-non-reserved-keywords.html
-
+  url: sql-migration-old.html
```
Comment (viirya) [inline docs/_data/menu-sql.yaml]: Do we need to keep these old links?

**Review by viirya — APPROVED** [review]
Thanks for doing this. It looks good.

**Comment by dongjoon-hyun** [comment]
Merged to master.
cc @jiangxb1987 since he is a release manager for 3.0.0-preview.
BTW, @jiangxb1987 . The broken migration tab sidebar as you see in this PR description is an existing issue in master branch.

**Comment by HyukjinKwon** [comment]
Thanks @dongjoon-hyun @viirya. I was on vacation so the reaction was late.

**Pull request closed without merging**